### PR TITLE
Feature/dm/marketplace license

### PR DESCRIPTION
**Pull request by dbmeneses**

### CODE REVIEW

File: server/sonar-server/src/main/java/org/sonar/server/plugins/PluginDownloader.java [hunk 1]
```diff
@@ -20,35 +20,150 @@
...
+  private void downloadRelease(Release release) throws URISyntaxException, IOException {
+    String url = release.getDownloadUrl();
+
+    URI uri = new URI(url);
+    if (url.startsWith("file:")) {
+      // used for tests
+      File file = toFile(uri.toURL());
+      copyFileToDirectory(file, downloadDir);
+    } else {
+      String filename = substringAfterLast(uri.getPath(), "/");
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/PluginDownloader.java]: lines 150 to 123 ought to be shared across the 2 download implementation.
This is support for updatecenter inconsistencies or for some "old" plugins. This is not download logic per say.

File: server/sonar-server/src/test/java/org/sonar/server/edition/ws/ApplyLicenseActionTest.java [hunk 2]
```diff
@@ -125,25 +142,109 @@ public void request_fails_with_BadRequestException_is_pendingStatus_is_not_NONE(
...
+
+  private void assertResponse(TestResponse response, String expectedNextEditionKey, String expectedEditionKey,
+    PendingStatus expectedPendingStatus) throws IOException {
+    StatusResponse parsedResponse = WsEditions.StatusResponse.parseFrom(response.getInputStream());
+    assertThat(parsedResponse.getCurrentEditionKey()).isEqualTo(expectedEditionKey);
+    assertThat(parsedResponse.getNextEditionKey()).isEqualTo(expectedNextEditionKey);
+    assertThat(parsedResponse.getInstallationStatus()).isEqualTo(WsEditions.InstallationStatus.valueOf(expectedPendingStatus.toString()));
+  }
+
+  private void setPendingLicense(PendingStatus pendingStatus) {
```
Comment (sns-seb) [inline server/sonar-server/src/test/java/org/sonar/server/edition/ws/ApplyLicenseActionTest.java]: hum... this works, but it's highly misleading as it mocks states which the production implementation can't get to.

I wonder if it would be safer to use the production implementation of `MutableEditionManagementState` in the test. A bit less convenient, but ensures relevant tests.

File: server/sonar-server/src/main/java/org/sonar/server/plugins/PluginDownloader.java [hunk 3]
```diff
@@ -20,35 +20,150 @@
...
+    try {
+      if (downloadDir.exists()) {
+        org.sonar.core.util.FileUtils.cleanDirectory(downloadDir);
+      }
+    } catch (IOException e) {
+      throw new IllegalStateException("Fail to clean the plugin downloads directory: " + downloadDir, e);
+    }
+  }
+
+  public boolean hasDownloads() {
```
Comment (sns-seb) [inline server/sonar-server/src/main/java/org/sonar/server/plugins/PluginDownloader.java]: not used. can be removed.

File: server/sonar-server/src/main/java/org/sonar/server/edition/License.java [hunk 4]
```diff
@@ -48,11 +59,31 @@ public String getEditionKey() {
...
+    try {
+      String data = new String(Base64.decodeBase64(base64.trim().getBytes(StandardCharsets.UTF_8)), StandardCharsets.UTF_8);
+
+      Properties props = new Properties();
+      props.load(new StringReader(data));
+
+      Collection<String> plugins = Arrays.asList(StringUtils.split(props.getProperty(PLUGINS_KEY), ','));
+      String editionKey = props.getProperty(EDITION_KEY);
+
+      if (editionKey != null && !plugins.isEmpty()) {
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/edition/License.java]: Indeed

File: server/sonar-server/src/main/java/org/sonar/server/edition/License.java [hunk 5]
```diff
@@ -48,11 +59,31 @@ public String getEditionKey() {
...
+        return Optional.of(new License(editionKey, plugins, base64));
+      }
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/edition/License.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/edition/License.java [hunk 6]
```diff
@@ -48,11 +59,31 @@ public String getEditionKey() {
...
+    } catch (Exception e) {
+      // do nothing
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/edition/License.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/edition/ws/PreviewAction.java [hunk 7]
```diff
@@ -71,10 +73,10 @@ public void handle(Request request, Response response) throws Exception {
       throw BadRequestException.create("Can't apply a license when applying one is already in progress");
     }
 
-    String license = request.mandatoryParam(PARAM_LICENSE);
-    License newLicense = new License(license, Collections.emptyList(), license);
-    NextState nextState = computeNextState(newLicense);
+    String licenseParam = request.mandatoryParam(PARAM_LICENSE);
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/edition/ws/PreviewAction.java]: Ok, but I would rephrase it as "The '%s' parameter is empty", otherwise it might also not be very clear

File: server/sonar-server/src/main/java/org/sonar/server/plugins/PluginDownloader.java (hunk 3 above)
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/plugins/PluginDownloader.java]: ok

File: server/sonar-server/src/test/java/org/sonar/server/edition/LicenseTest.java [hunk 8]
```diff
@@ -71,8 +76,56 @@ public void constructor_fails_if_content_is_empty() {
   }
 
   @Test
+  public void parse_is_empty_if_license_is_invalid_string() {
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/test/java/org/sonar/server/edition/LicenseTest.java]: ok for both except that I disagree about the need for a random string knowing what we are testing.

File: server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionInstaller.java [hunk 9]
```diff
@@ -57,21 +72,24 @@ public void install(Set<String> editionPlugins) {
...
   }
 
-  private void asyncInstall(Set<String> editionPluginKeys) {
+  private void asyncInstall(Set<String> editionPluginKeys, UpdateCenter updateCenter) {
     try {
-      // TODO clean previously staged edition installations, or fail?
-      // editionPluginDownloader.cancelDownloads();
-      // editionPluginUninstaller.cancelUninstalls();
-      editionPluginDownloader.installEdition(pluginsToInstall(editionPluginKeys));
+      editionPluginDownloader.installEdition(pluginsToInstall(editionPluginKeys), updateCenter);
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionInstaller.java]: Ok

File: server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionInstaller.java [hunk 10]
```diff
@@ -19,35 +19,50 @@
...
+    this.updateCenterMatrixFactory = updateCenterMatrixFactory;
+    this.editionManagementState = editionManagementState;
   }
 
-  public void install(Set<String> editionPlugins) {
+  /**
+   * Refreshes the update center, and submits in a executor a task to download all the needed plugins (asynchronously).
+   * If the update center is disabled or if we are offline, the task is not submitted and false is returned. 
+   * @return true if a task was submitted to perform the download, false if update center is unavailable.
+   */
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionInstaller.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/edition/ws/PreviewAction.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (sns-seb) [reply server/sonar-server/src/main/java/org/sonar/server/edition/ws/PreviewAction.java]: "empty" is ok, but I didn't find it necessary.
when I see a URL with "foo=" I don't see the difference with not providing the parameter at all

File: server/sonar-server/src/main/java/org/sonar/server/edition/License.java [hunk 11]
```diff
@@ -19,22 +19,33 @@
...
 import javax.annotation.concurrent.Immutable;
+import org.apache.commons.codec.binary.Base64;
+import org.apache.commons.lang.StringUtils;
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.base.Preconditions.checkNotNull;
 
 @Immutable
 public class License {
+  public static final String EDITION_KEY = "Edition";
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/edition/License.java]: ok

File: server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionPluginDownloader.java [hunk 12]
```diff
@@ -19,41 +19,88 @@
...
+  private final HttpDownloader downloader;
+
+  public EditionPluginDownloader(HttpDownloader downloader, ServerFileSystem fileSystem) {
+    this.downloadDir = fileSystem.getEditionDownloadedPluginsDir().toPath();
+    this.downloader = downloader;
+    this.tmpDir = downloadDir.resolveSibling(downloadDir.getFileName() + "_tmp");
   }
 
-  public void installEdition(Set<String> pluginKeys) {
+  public void installEdition(Set<String> pluginKeys, UpdateCenter updateCenter) {
```
[the remark this answers is not in the document]
Comment (dbmeneses) [reply server/sonar-server/src/main/java/org/sonar/server/plugins/edition/EditionPluginDownloader.java]: ok

File: server/sonar-server/src/test/java/org/sonar/server/edition/LicenseTest.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (sns-seb) [reply server/sonar-server/src/test/java/org/sonar/server/edition/LicenseTest.java]: I tend to like the use of a random string for the reason that if you're in the TDD mindset, then the code could very well fail only for that constant string :)
